src/coreclr/jit/optimizebools.cpp

```diff
@@ -1570,3 +1570,698 @@ PhaseStatus Compiler::optOptimizeBools()
+}
+
+//-----------------------------------------------------------------------------
+//  optFindSpecificPattern: Optimize range check for if (A || B || C || D) pattern.
+//
+//  Returns:
+//      MODIFIED_NOTHING if no optimization is performed.
+//      MODIFIED_EVERYTHING otherwise.
+//
+//  Notes:
+//      Detect if (a == val1 || a == val2 || a == val3 || a == val4) pattern and change it to switch tree to reduce
+//      jumps and perform bit operation instead.
+//
+PhaseStatus Compiler::optFindSpecificPattern()
+{
+#ifdef DEBUG
+    if (this->verbose)
+    {
+        printf("\n*************** In optFindSpecificPattern()\n");
+    }
+#endif // DEBUG
+
+    if (!opts.OptimizationEnabled())
+    {
+        return PhaseStatus::MODIFIED_NOTHING;
+    }
+
+    OptRangePatternDsc optRngPattern(this);
+
+    bool         printed           = false;
+    int          patternIndex      = 0;       // The index of the pattern in the array
+    bool         foundPattern      = false;
+    unsigned int firstPatternBBNum = 0;       // Basic block number of the first pattern found
+    BasicBlock*  prevBb            = fgFirstBB;
+
+    if (fgFirstBB->bbNext == nullptr)
+    {
+        return PhaseStatus::MODIFIED_NOTHING;
+    }
+
+    for (BasicBlock* currBb = fgFirstBB->bbNext; currBb != nullptr; currBb = currBb->bbNext)
+    {
+        //if ((currBb->KindIs(BBJ_COND) || currBb->KindIs(BBJ_RETURN)) && prevBb != nullptr && prevBb->KindIs(BBJ_COND))
+        if (currBb->KindIs(BBJ_COND) && prevBb != nullptr && prevBb->KindIs(BBJ_COND))
+        {
+#ifdef DEBUG
+            if (this->verbose)
+            {
+                if (!printed)
+                {
+                    printf("Display Basic Blocks for optFindSpecificPattern:\n");
+                    fgDispBasicBlocks(true);
+                    printed = true;
+                }
+            }
+#endif // DEBUG
+
+            if (currBb->lastStmt() == currBb->firstStmt() && prevBb->lastStmt() == prevBb->firstStmt())
+            {
+                auto condition1 = currBb->lastStmt()->GetRootNode()->gtGetOp1();
+                auto condition2 = prevBb->lastStmt()->GetRootNode()->gtGetOp1();
+                if (condition1->OperIsCompare() && condition2->OperIsCompare())
+                {
+                    if (currBb->bbJumpDest == prevBb->bbJumpDest ||
+                        (condition1->OperIs(GT_NE) && condition2->OperIs(GT_EQ) && currBb->bbNext == prevBb->bbJumpDest))
+                    {
+                        // Check both conditions to have constant on the right side
+                        if (condition1->gtGetOp2()->IsIntegralConst() && condition2->gtGetOp2()->IsIntegralConst())
+                        {
+                            auto leftCondition1 = condition1->gtGetOp1(); // op1 of condition1 from currBb
+                            auto leftCondition2 = condition2->gtGetOp1(); // op1 of condition2 from prevBb
+                            if (leftCondition1->IsLocal() &&
```

Comment:
I don't think this is a correct check - better make it more conservative and check `OperIs(GT_LCL_VAR)`
because in your case a struct field can pass this check but two locals will have different offsets